### PR TITLE
dbscan implement

### PR DESCRIPTION
**Pull request by DoubleCorner**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
+  "description": "dbscan",
   "main": "./lib/index.js",
   "module": "./lib-esm/index.js",
   "types": "./lib/index.d.ts",
-  "keywords": [],
+  "keywords": [
+    "clustering",
+    "dbscan"
+  ],
   "author": "Luc Patiny",
```
Comment (lpatiny) [inline package.json]: Please put your name here.

File: README.md [hunk 2]
```diff
@@ -14,10 +14,19 @@
 ## Usage
```
Comment (lpatiny) [inline README.md:16]: Are you able to improve the README file with explanation of the algorithm, link to wikipedia (https://en.wikipedia.org/wiki/DBSCAN) and/or link to publications (using DOI !).

File: src/index.ts [hunk 3]
```diff
@@ -1,7 +1,79 @@
...
+}
+function rangeQuery(
+  cur: IPoint,
+  data: IPoint[],
+  epsilon: number,
+  distance: (p: number[], q: number[]) => number,
+) {
+  return data.filter((point) => distance(point.value, cur.value) <= epsilon);
+}
+export function dbscan(
```
Comment (lpatiny) [inline src/index.ts]: Would it be possible to add jsDoc (well tsdoc in fact) for this function ? We are building automatically a 'userguide' using a GitHub action like for this project for example: https://mljs.github.io/spectra-processing/

File: src/index.ts [hunk 4]
```diff
@@ -1,7 +1,79 @@
...
+  points: number[][],
+  epsilon: number = 1,
+  minPoints: number = 2,
+  distance: (p: number[], q: number[]) => number = euclidean,
+) {
+  if (!(points instanceof Array)) {
```
Comment (lpatiny) [inline src/index.ts:34]: When methods have optional parameters we like to put an 'options' object with default value '{}' and then to deconstruct it.

Like in this example https://github.com/mljs/spectra-processing/blob/953886481787535ece7ddcd6d60d9301ef61ec6a/src/xy/xyFilterX.ts#L5-L33

File: src/index.ts [hunk 5]
```diff
@@ -1,7 +1,79 @@
...
  */
-export function myModule(): number {
-  return 42;
+interface IPoint {
+  index: number;
+  value: number[];
+  label: number;
+}
+function rangeQuery(
+  cur: IPoint,
```
Comment (lpatiny) [inline src/index.ts]: In general we avoid abbreviation so that the code is relatively easy to read. I guess cur is `current`

File: src/index.ts [hunk 6]
```diff
@@ -1,7 +1,79 @@
...
+    throw Error('points must be of type array, ' + typeof points + ' given');
+  }
+  let clusterId = 0;
+  const data: IPoint[] = points.map((point, index) => ({
+    index,
+    value: point,
+    label: -1,
+  }));
+
+  data.forEach((point) => {
```
Comment (lpatiny) [inline src/index.ts]: When making algorithm we should avoid forEach that is quite slow (and a little bit more difficult to read). You may have a look at this kind of benchmark: https://jsbench.me/phkg20iq7z

Please use a `for (const point of data)`

File: src/index.ts [hunk 7]
```diff
@@ -1,7 +1,79 @@
...
+  cur: IPoint,
+  data: IPoint[],
+  epsilon: number,
+  distance: (p: number[], q: number[]) => number,
+) {
+  return data.filter((point) => distance(point.value, cur.value) <= epsilon);
+}
+export function dbscan(
+  points: number[][],
+  epsilon: number = 1,
```
Comment (lpatiny) [inline src/index.ts]: jsDoc of all the parameter would help. The minPoints = 2 as default value is coming from literature ?

File: src/index.ts [hunk 8]
```diff
@@ -1,7 +1,79 @@
...
+  minPoints: number = 2,
+  distance: (p: number[], q: number[]) => number = euclidean,
+) {
+  if (!(points instanceof Array)) {
+    throw Error('points must be of type array, ' + typeof points + ' given');
+  }
+  let clusterId = 0;
+  const data: IPoint[] = points.map((point, index) => ({
```
Comment (lpatiny) [inline src/index.ts]: I think using a traditional loop with 'push' would be much faster: https://jsbench.me/phkg20iq7z

**Review by lpatiny — CHANGES_REQUESTED** [review]
Thanks for your contribution ! It is great to see that people are contributing to our projects.

I did some comments on the code mainly little tricks for faster code and improve the visibility of the package once published.

Not sure if the points dimension are always '2'. If yes may this could appear in the typescript definition: https://stackoverflow.com/questions/41139763/how-to-declare-a-fixed-length-array-in-typescript

File: src/index.ts (hunk 7 above)
Comment (DoubleCorner) [reply src/index.ts:21]: No, only at least 2 points can form a cluster

File: README.md (hunk 2 above)
Comment (DoubleCorner) [reply README.md:16]: ok

File: src/index.ts (hunk 3 above)
Comment (DoubleCorner) [reply src/index.ts]: I do not know how to write it, but I wite it in ReadMe.md

File: src/index.ts (hunk 4 above)
Comment (DoubleCorner) [reply src/index.ts:34]: ok

File: src/index.ts (hunk 5 above)
Comment (DoubleCorner) [reply src/index.ts]: ok

File: src/index.ts (hunk 6 above)
Comment (DoubleCorner) [reply src/index.ts]: ok

File: src/index.ts (hunk 7 above)
Comment (DoubleCorner) [reply src/index.ts]: No, because at least two points can form clusters

File: src/index.ts (hunk 8 above)
Comment (DoubleCorner) [reply src/index.ts]: ok

**Pull request merged**